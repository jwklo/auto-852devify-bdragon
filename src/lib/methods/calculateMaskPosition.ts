import * as faceapi from 'face-api.js';
import { MaskDimension } from '../types';
import { maskTypes } from '@/settings/global';


/**
 * Description placeholder
 * @date 2023/10/26 - 上午12:02:40
 *
 * @param {faceapi.Point[]} points
 * @returns {{ x: number; y: number; }}
 */
function middlePoint(points: faceapi.Point[]) {
    const x = points.reduce((acc, curr) => acc + curr.x, 0) / points.length;
    const y = points.reduce((acc, curr) => acc + curr.y, 0) / points.length;
    return { x, y };
}

/**
 * Get the mask image dimension from the HTMLImageElement
 * @date 2023/10/26 - 上午12:03:12
 *
 * @export
 * @param {HTMLImageElement} mask
 * @param {number} [widthAdjust=1]
 * @returns {MaskDimension}
 */
export function getMaskDimension(mask: HTMLImageElement, widthAdjust: number = 1): MaskDimension {
    const width = mask.width;
    const height = mask.height;
    return { width, height, widthAdjust }
}


/**
 * Calculate the positions of mask in the background image
 * @date 2023/10/26 - 上午12:03:12
 *
 * @export
 * @param {number} maskType
 * @param {faceapi.FaceLandmarks68} landmarks
 * @param {MaskDimension} maskDimension
 * @param {boolean} flipMask
 * @returns {Array<number>}
 * @example [midX, midY, width, height, angle, hscale]
 * @description midX,midY: Middle point of mask image in the background image
 * @description width,height: width/height of mask image
 * @description angle, hscale: rotated angle and flipping flag of the mask image
 */

export function calcuateMaskPosition(maskType: number, landmarks: faceapi.FaceLandmarks68, maskDimension: MaskDimension, flipMask: boolean): Array<number> {
    //Eyes Middle point
    const leftEyeMid = middlePoint(landmarks.getLeftEye()); 
    const rightEyeMid = middlePoint(landmarks.getRightEye());

    //Most left/right points of eyes/face
    const leftEye = landmarks.positions[36];
    const rightEye = landmarks.positions[45];
    const leftFace = landmarks.positions[0];
    const rightFace = landmarks.positions[16];
    const leftEyebrow = landmarks.positions[19];
    const rightEyebrow = landmarks.positions[24];

    //face width
    const faceWidth = leftFace.x - rightFace.x;

    //Mid points between face and eyes boundary
    const leftFaceMid = { x: (leftFace.x + leftEye.x) / 2, y: (leftFace.y + leftEye.y) / 2 };
    const rightFaceMid = { x: (rightFace.x + rightEye.x) / 2, y: (rightFace.y + rightEye.y) / 2 };

    //Face height
    const faceLeftTop = landmarks.positions[19];
    const faceRightTop = landmarks.positions[24];
    const faceBottom = landmarks.positions[8];
    const [heightLeftA, heightLeftB] = [(faceLeftTop.x - faceBottom.x), (faceLeftTop.y - faceBottom.y)];
    const [heightRightA, heightRightB] = [(faceRightTop.x - faceBottom.x), (faceRightTop.y - faceBottom.y)];
    const leftFaceHeight = Math.sqrt(heightLeftA * heightLeftA + heightLeftB * heightLeftB);
    const rightFaceHeight = Math.sqrt(heightRightA * heightRightA + heightRightB * heightRightB);

    //Take the longer one as the face height
    const faceHeight = Math.max(leftFaceHeight, rightFaceHeight);

    //Nose position
    const nose = landmarks.positions[27] //27 or 28

    //Length of left/right between left/right face and nose. To detect the direction of the face
    const [leftA, leftB] = [(nose.x - leftFace.x), (nose.y - leftFace.y)];
    const [rightA, rightB] = [(nose.x - rightFace.x), (nose.y - rightFace.y)];
    const leftFaceDistance = Math.sqrt(leftA * leftA + leftB * leftB); //a2 + b2 = c2 ,Pythagorean theorem
    const rightFaceDistance = Math.sqrt(rightA * rightA + rightB * rightB);

    //for flipping according to the face direction
    const hscale = rightFaceDistance > leftFaceDistance && flipMask ? -1 : 1;
    //TOBEREMOVE: console.log("face:", leftEye, leftFace, rightEye, rightFace, leftFaceDistance, rightFaceDistance, leftFaceMid, rightFaceMid);

    //Calculation the attributes accoring to maskType
    switch (maskType) {
        //Base on the middle point of the eyes
        case maskTypes.byEyesMiddle:
            return (() => {
                const dx = rightEyeMid.x - leftEyeMid.x;
                const dy = rightEyeMid.y - leftEyeMid.y;
                const angle = Math.atan2(dy, dx);
                const width = Math.abs(rightEyeMid.x - leftEyeMid.x) * maskDimension.widthAdjust;
                const height = Math.abs(width * maskDimension.height / maskDimension.width);
                const midX = (rightEyeMid.x + leftEyeMid.x) / 2;
                const midY = (rightEyeMid.y + leftEyeMid.y) / 2;
                return [midX, midY, width, height, angle, hscale]
            })();

        //Base on the face width 
        case maskTypes.byFaceWidth:
            return (() => {
                const dx = rightFace.x - leftFace.x;
                const dy = rightFace.y - leftFace.y;
                const angle = Math.atan2(dy, dx);
                const width = Math.abs(faceWidth) * maskDimension.widthAdjust;
                const height = Math.abs(width * maskDimension.height / maskDimension.width);
                const midX = nose.x;
                const midY = nose.y;
                return [midX, midY, width, height, angle, hscale]
            })();

        //Mask the whole face without concern the ratio
        case maskTypes.fullFace: //Not by ratio
            return (() => {
                const dx = rightFace.x - leftFace.x;
                const dy = rightFace.y - leftFace.y;
                const angle = Math.atan2(dy, dx);
                const width = Math.abs(faceWidth) * maskDimension.widthAdjust;

                const height = faceHeight * 1.5;
                const midX = nose.x;
                const midY = nose.y;
                return [midX, midY, width, height, angle, hscale]
            })();
        case maskTypes.overTheHead:
            return (() => {
                const dx = rightFace.x - leftFace.x;
                const dy = rightFace.y - leftFace.y;
                const angle = Math.atan2(dy, dx);
                const width = Math.abs(faceWidth) * maskDimension.widthAdjust * 1.2;
                const height = Math.abs(width * maskDimension.height / maskDimension.width);
                const midX = nose.x;
                const midY = Math.max(leftEyebrow.y, rightEyebrow.y) - (height);
                return [midX, midY, width, height, angle, hscale]
            })();

        //Type 0: Base on middle of the face boundary and the eye boundary
        default:
            return (() => {
               
                
                const width = Math.abs(leftFaceMid.x - rightFaceMid.x) * maskDimension.widthAdjust;
                const height = Math.abs(width * maskDimension.height / maskDimension.width);
                const dx = rightFaceMid.x - leftFaceMid.x;
                const dy = rightFaceMid.y - leftFaceMid.y;
                const angle = Math.atan2(dy, dx);
                const midX = nose.x;
                const midY = nose.y;
                return [midX, midY, width, height, angle, hscale]
            })();

    }


}