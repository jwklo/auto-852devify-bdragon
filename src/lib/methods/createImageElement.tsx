/**
 * Create image from source uri
 * @date 2023/10/26 - 上午12:25:20
 *
 * @param {string} src
 * @returns {Promise<HTMLImageElement>}
 */
function createImageElement(src: string): Promise<HTMLImageElement> {
    return new Promise((resolve, reject) => {
        const img = document.createElement('img');
        img.src = src;
        img.onload = () => resolve(img);
        img.onerror = reject;
    });
}

export { createImageElement }