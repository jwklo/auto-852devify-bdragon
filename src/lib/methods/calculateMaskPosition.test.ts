/**
 * @jest-environment jsdom
 */

//import { MaskDimension } from "../types";
import { getMaskDimension, calcuateMaskPosition } from "./calculateMaskPosition";
import { expect, jest, test, describe, it } from '@jest/globals';
import { createImageElement } from "./createImageElement";
import * as faceapi from 'face-api.js';

function sampleLanmarks(){
    return {
        dimension: {w: 51,h: 72}, 
        shift: {x: 654.0684080123901,y: 156.5300187579656},
        lanmarks: [
        {
            "_x": 0.03270969167351723,
            "_y": 0.3386686146259308
        },
        {
            "_x": 0.01784905605018139,
            "_y": 0.4266079366207123
        },
        {
            "_x": 0.021425720304250717,
            "_y": 0.5130588412284851
        },
        {
            "_x": 0.03573403134942055,
            "_y": 0.588279664516449
        },
        {
            "_x": 0.05775056779384613,
            "_y": 0.6829872727394104
        },
        {
            "_x": 0.1033591702580452,
            "_y": 0.7633322477340698
        },
        {
            "_x": 0.15137328207492828,
            "_y": 0.8277785778045654
        },
        {
            "_x": 0.21796981990337372,
            "_y": 0.8992071747779846
        },
        {
            "_x": 0.35328495502471924,
            "_y": 0.9579337239265442
        },
        {
            "_x": 0.5140458345413208,
            "_y": 0.9465249180793762
        },
        {
            "_x": 0.6411281228065491,
            "_y": 0.9032599925994873
        },
        {
            "_x": 0.7421179413795471,
            "_y": 0.8601421117782593
        },
        {
            "_x": 0.8257849216461182,
            "_y": 0.7922584414482117
        },
        {
            "_x": 0.8901783227920532,
            "_y": 0.7078917622566223
        },
        {
            "_x": 0.9258625507354736,
            "_y": 0.6258351802825928
        },
        {
            "_x": 0.9638902544975281,
            "_y": 0.5421010255813599
        },
        {
            "_x": 0.9898282885551453,
            "_y": 0.45283395051956177
        },
        {
            "_x": 0.10680602490901947,
            "_y": 0.2391723394393921
        },
        {
            "_x": 0.15451295673847198,
            "_y": 0.21373251080513
        },
        {
            "_x": 0.215469092130661,
            "_y": 0.21567551791667938
        },
        {
            "_x": 0.2749539911746979,
            "_y": 0.23191803693771362
        },
        {
            "_x": 0.3292182683944702,
            "_y": 0.25665339827537537
        },
        {
            "_x": 0.5789802670478821,
            "_y": 0.2862688899040222
        },
        {
            "_x": 0.6441187858581543,
            "_y": 0.2766788899898529
        },
        {
            "_x": 0.7222675085067749,
            "_y": 0.27767953276634216
        },
        {
            "_x": 0.8036983013153076,
            "_y": 0.2969423830509186
        },
        {
            "_x": 0.8612868785858154,
            "_y": 0.3397844135761261
        },
        {
            "_x": 0.4233970642089844,
            "_y": 0.38804349303245544
        },
        {
            "_x": 0.3962896466255188,
            "_y": 0.45450738072395325
        },
        {
            "_x": 0.37126627564430237,
            "_y": 0.5118997097015381
        },
        {
            "_x": 0.35344114899635315,
            "_y": 0.5641416907310486
        },
        {
            "_x": 0.29961544275283813,
            "_y": 0.5809475183486938
        },
        {
            "_x": 0.3256005048751831,
            "_y": 0.5982100367546082
        },
        {
            "_x": 0.37126150727272034,
            "_y": 0.6110795140266418
        },
        {
            "_x": 0.4245169460773468,
            "_y": 0.608782172203064
        },
        {
            "_x": 0.46896156668663025,
            "_y": 0.6093466281890869
        },
        {
            "_x": 0.1732344925403595,
            "_y": 0.33399227261543274
        },
        {
            "_x": 0.20742549002170563,
            "_y": 0.3289088010787964
        },
        {
            "_x": 0.27123770117759705,
            "_y": 0.33651289343833923
        },
        {
            "_x": 0.32247206568717957,
            "_y": 0.3668733239173889
        },
        {
            "_x": 0.2718980014324188,
            "_y": 0.373725950717926
        },
        {
            "_x": 0.20865866541862488,
            "_y": 0.3601033389568329
        },
        {
            "_x": 0.5733473896980286,
            "_y": 0.39970916509628296
        },
        {
            "_x": 0.6333433389663696,
            "_y": 0.382898211479187
        },
        {
            "_x": 0.6996922492980957,
            "_y": 0.3909933865070343
        },
        {
            "_x": 0.7436284422874451,
            "_y": 0.4151311218738556
        },
        {
            "_x": 0.6920861005783081,
            "_y": 0.42917490005493164
        },
        {
            "_x": 0.623864471912384,
            "_y": 0.4173659384250641
        },
        {
            "_x": 0.19722861051559448,
            "_y": 0.6685444712638855
        },
        {
            "_x": 0.2425702065229416,
            "_y": 0.6526060104370117
        },
        {
            "_x": 0.32308170199394226,
            "_y": 0.6543692350387573
        },
        {
            "_x": 0.36805886030197144,
            "_y": 0.6639478802680969
        },
        {
            "_x": 0.4167270064353943,
            "_y": 0.6634411811828613
        },
        {
            "_x": 0.5281736254692078,
            "_y": 0.6943700909614563
        },
        {
            "_x": 0.6172153949737549,
            "_y": 0.727498471736908
        },
        {
            "_x": 0.5097690224647522,
            "_y": 0.7839610576629639
        },
        {
            "_x": 0.42670363187789917,
            "_y": 0.8055270314216614
        },
        {
            "_x": 0.35006627440452576,
            "_y": 0.8047371506690979
        },
        {
            "_x": 0.2895500659942627,
            "_y": 0.7879573702812195
        },
        {
            "_x": 0.23603293299674988,
            "_y": 0.7502620816230774
        },
        {
            "_x": 0.20724175870418549,
            "_y": 0.6727645397186279
        },
        {
            "_x": 0.30541783571243286,
            "_y": 0.6783864498138428
        },
        {
            "_x": 0.36865919828414917,
            "_y": 0.6886245012283325
        },
        {
            "_x": 0.43727537989616394,
            "_y": 0.6996099948883057
        },
        {
            "_x": 0.6002424359321594,
            "_y": 0.7267441153526306
        },
        {
            "_x": 0.429147869348526,
            "_y": 0.7603375315666199
        },
        {
            "_x": 0.3621083199977875,
            "_y": 0.7591655850410461
        },
        {
            "_x": 0.3006831109523773,
            "_y": 0.7460619211196899
        }
    ]}
}


it('Get Dimension of the image (getMaskDimension)', async () => {
    //expect.assertions(1);

    let imageElement = await createImageElement('http://852dev.bdragon.xyz/red160px.png');
    const md = getMaskDimension(imageElement);
    console.log(imageElement,md);

    expect(md).toMatchSnapshot();
})

it('Get masks positions on the background image (calcuateMaskPosition)', async () => {
    //expect.assertions(1);
    let points:faceapi.Point[] = []; 
    let {dimension, shift, lanmarks}= sampleLanmarks();
    let dim = new faceapi.Dimensions(dimension.w,dimension.h)
    lanmarks.forEach(lm => {
        points.push(new faceapi.Point(lm._x, lm._y));
    })

  
    let shiftPoint = new faceapi.Point(shift.x,shift.y);
    //let shift = new faceapi.Point(0,0);
   
    let landmarks = new faceapi.FaceLandmarks68(points, dim, shiftPoint);
    let maskDimension = {width: 160, height: 60, widthAdjust: 1};
    let flipMask = true;

    let maskPosition = calcuateMaskPosition(0, landmarks, maskDimension, flipMask)
    console.log("maskPosition", points, landmarks, maskPosition);
    expect(maskPosition).toMatchSnapshot();
})

