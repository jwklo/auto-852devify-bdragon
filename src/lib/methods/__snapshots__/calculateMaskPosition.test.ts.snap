// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get Dimension of the image (getMaskDimension) 1`] = `
{
  "height": 60,
  "width": 160,
  "widthAdjust": 1,
}
`;

exports[`Get masks positions on the background image (calcuateMaskPosition) 1`] = `
[
  675.6616582870483,
  184.46915025630238,
  38.9515699390322,
  14.606838727137074,
  0.1785819852982527,
  -1,
]
`;
